language-support/ts/codegen/tests/ts/daml-json-types/src/index.ts

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+import * as jtv from '@mojotech/json-type-validation';
+
+/**
+ * Interface for companion objects of serializable types. Its main purpose is
+ * to describe the JSON encoding of values of the serializable type.
+ */
+export interface Serializable<T> {
+  // NOTE(MH): This must be a function to allow for mutually recursive decoders.
+  decoder: () => jtv.Decoder<T>;
+}
+
+/**
+ * Identifier of a DAML template.
+ */
+export type TemplateId = {
+  packageId: string;
+  moduleName: string;
+  entityName: string;
+}
+
+/**
+ * Companion object of the `TemplateId` type.
+ */
+const TemplateId: Serializable<TemplateId> = {
+  decoder: () => jtv.object({
+    packageId: jtv.string(),
+    moduleName: jtv.string(),
+    entityName: jtv.string(),
+  })
+}
+
+/**
+ * Interface for objects representing DAML templates. It is similar to the
+ * `Template` type class in DAML.
+ */
+export interface Template<T extends {}> extends Serializable<T> {
+  templateId: TemplateId;
+}
+
+/**
+ * Interface for objects representing DAML choices. It is similar to the
+ * `Choice` type class in DAML.
+ */
+export interface Choice<T, C> extends Serializable<C> {
+  template: Template<T>;
+  choiceName: string;
+}
+
+/**
+ * Interface for objects representing DAML templates with an `Archive` choice.
+ */
+export interface ArchivableTemplate<T extends {}> extends Template<T> {
+  Archive: Choice<T, {}>;
+}
+
+/**
+ * The counterpart of DAML's `()` type.
+ */
+export type Unit = {};
+
+/**
+ * Companion obect of the `Unit` type.
+ */
+export const Unit: Serializable<Unit> = {
+  decoder: () => jtv.object({}),
+}
+
+/**
+ * The counterpart of DAML's `Bool` type.
+ */
+export type Bool = boolean;
+
+/**
+ * Companion object of the `Bool` type.
+ */
+export const Bool: Serializable<Bool> = {
+  decoder: jtv.boolean,
+}
+
+/**
+ * The counterpart of DAML's `Int` type. We represent `Int`s as string in order
+ * to avoid a loss of precision.
+ */
+export type Int = string;
```

Comment:
We should switch this to some kind of type that lets you do arithmetic at some point. Otherwise the first thing users are going to do is to convert this to a JS number which defeats the point.

Reply:
I agree. For now, this is "good enough" though.